rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+---
+feature: nix-encryption
+start-date: 2016-03-28
+author: Eelco Dolstra
+co-authors:
+related-issues: https://github.com/NixOS/nix/issues/8, https://github.com/NixOS/nix/pull/329
+---
+
+# Summary
+[summary]: #summary
+
+We currently lack a way to store secret information in the Nix
+store. The proposal is to add add a builtin function to Nix to allow
+secrets to be encrypted at evaluation time with a key. At runtime,
+files containing encrypted data can be decrypted using the same key.
+
+# Motivation
+[motivation]: #motivation
+
+It is often necessary to store secrets such as passwords or SSL
+certificates in NixOS system closures. Unfortunately, the Nix store is
+world-readable, so this is currently not possible in a secure way;
+anybody who gains access to the filesystem of a NixOS machine can read
+any file in its Nix store. Nevertheless, lots of NixOS modules *do*
+store secrets in the Nix store (see
+https://github.com/NixOS/nixpkgs/issues/24288), many of them lacking
+warnings about the danger of doing so.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposal consists of the following components:
+
+* A builtin function `encryptString :: Path -> String -> String` that
+  encrypts a string using a symmetric key stored in the specified
+  path. The encryption is done using libsodium's authenticated
+  encryption function `crypto_secretbox_easy()`. The resulting string
+  has the format `<{|nixcrypt:base64-encoded-data|}>`, where
+  `base64-encoded-data` is the base-64-encoded output of
+  `crypto_secretbox_easy()`, along with the nonce used.
+
+* A command `nix decrypt` to decrypt files containing encrypted
+  strings produced by `encryptString`. This command searches for
+  `<{|nixcrypt:...|}>` fragments and decrypts them using a decryption
+  key specified on the command line. Any other data is copied
+  verbatim.
+
+* A command `nix generate-key` to generate a key in the format
+  expected by `encryptString` and `nix decrypt`.
+
+A typical usage in a module NixOS would look like this:
+
+```
+let
+  configFile = pkgs.writeText "foo.conf"
+    ''
+      # Store the password of the foo service in encrypted form.
+      password=${builtins.encryptString <nixos-store-key> cfg.password}
```

Comment:
I'm aware of two solutions to providing the password:

* Approach 1: The user does not supply a secret at all and the derivation auto-generates the secret at startup.  The secret is generated so that only `root` can access it and the user has to `ssh` into the machine as `root` to obtain the credential if they need it to log in.
* Approach 2: The user logs into the machine after deploy and deposits the secret in a preconfigured location by the user after deploy and the program/service waits or fails until the secret is available.